IPIP-431: Opt-in Extensible CAR Metadata on Trustless Gateway (#431)
ipfs/specs: src/http-gateways/trustless-gateway.md

the raw data is already present in the parent block that links to the identity
CID.

## `meta` (content type parameter)

The `meta` parameter allows clients to request the server to include additional metadata about the CAR along with the response body.

The value of this parameter includes both the location where the metadata is given (e.g. `eof`) as well as the type of data received (e.g. `json`) separated by a `+`, to give a value such as `meta=eof+json`

When the location parameter is set to `eof`, which is currently the only supported value, the server SHOULD respond with the following response body:

```
<Response body as CARv1 stream> <0x00 byte> <Metadata>

> ```

Author: It looks like the template used by the website cannot handle `<>` characters correctly :( Is it something worth fixing or should you change this PR to use different characters?

E.g. follow the IPLD style:

```
Response body as CARv1 stream | 0x00 byt | Metadata
```

Author: <img width="831" alt="Screenshot 2023-10-18 at 16 16 38" src="https://github.com/ipfs/specs/assets/1140553/a3c2ec40-771b-4e66-877d-07e9b972990f">

Reviewer: This is a bug in website generator, not blocking this PR. 
Fix tracked in https://github.com/ipfs/spec-generator/issues/19

Reviewer: For now just use 
1```suggestion
```
[Response body as CARv1 stream] [0x00 byte] [Metadata]
```
````

Author: I see that https://github.com/ipfs/spec-generator/issues/19 was fixed, I will stick to `<>`.

The only supported value for the data type parameter is `json`. This signifies that the metadata MUST be a JSON object.

This parameter MUST only be used with CAR `version=1`.

When the parameter is not set or does not equal `eof+json`, the server SHOULD not add any extra blocks to the response, neither the 0x00 byte nor any metadata.

> When `meta=eof+json`, the JSON object SHOULD conform to the following [JSON schema](https://json-schema.org/).

Author: Discussion points:

- In the current spec & IPIP, we are formatting metadata as JSON. Should we say DAG-JSON instead?
- Do we want to serialise the metadata as a CAR block, prefixing the JSON data with `varint | CID` header?

Reviewer: @rvagg  thoughts? Value added in DAG-JSON prefixed with own CID is that it allows client to detect truncation beyond `0x00` byte.

Author: I believe clients can already easily detect truncation of the metadata block.
- The block is a DAG-JSON object, it must start with `{` and end with a matching `}`.
- If the block is truncated, it will not end with the matching `}` and the JSON parser will throw an error.

```json
{
  "type": "object",
  "properties": {
    "data": {
      "type": "object",
      "description": "Properties of the response"

>     },

Author: Discussion point:

In the current proposal, the top-level "data" object combines fields about "what was requested" (e.g. CAR & DAG params) with "what was returned" (e.g. CARv1 length in bytes).

I'd like to discuss an alternative: split `data` into two fields `req` and `res`. The first will describe what the client requested, the second will describe what the server returned.

Such division would allow us to shorten field names, e.g. `data.car_params.dup` can become `req.dups`.

Reviewer: Splitting into `req` and `res` sgtm, improves clarity

    "error": {
      "type": "string",
      "description": "Error message"
    },

>     "sig": {
>       "type": "string",
>       "description": "A signature, using the server's Ed2559 identity, over the `data` object serialized as JSON."

Reviewer: @bajtos 

- HTTP Gateways have no concept of "sever ED25519" introduced here. How one verifies the signature without knowing the pubkey?
  - One way to avoid being prescriptive about key type or its location, is to have `sig_key` with CID-encoded public  `libp2p-key` that can be used for signature verification.
    - The nice thing about this is that Gateway/client implementation will already have relevant code/library as we use these in IPNS and libp2p.
- If you sign JSON, you want it to be deterministic variant like DAG-JSON, otherwise someone will run into bugs when they use less strict JSON library in different languages.

```suggestion
   "sig_pubkey": {
      "type": "string",
      "description": "A libp2p-key used for signing"
    },
    "sig": {
      "type": "string",
      "description": "A signature, using the `sig_pubkey`, over the `data` object serialized as DAG-JSON."
```

Author: Here is our use case:
1. An untrusted/permissionless client makes a retrieval request to the Storage Provider's booster-http address advertised in IPNI.
2. The client submits the measurement to the SPARK orchestration layer.
3. Later, SPARK's evaluation service wants to verify that the client contacted the SP.

To do so, we must not accept signatures from any identity, only the signature from the identity advertised by SP to IPNI.

I am arguing this is true for everybody else who wants to use the signature to verify that a metadata block submitted by an untrusted party was indeed produced by the expected Trustless Gateway.

Consider a simple attack vector: the attacker takes the metadata block produced by the origin gateway and replaces the signature with one created using the attacker's identity. Clients verifying the signature against the `sig_pubkey` field in the metadata will not notice the attack.

Now I can see how including `sig_pubkey` can simplify troubleshooting:
- If `sig_pubkey` does not match the pubkey we expected, then we know the metadata block was signed by somebody else
- If `sig_pubkey` matches but the signature does not, then we know the metadata block was modified from the original.
Compare that with my proposal:
- If the signature is not valid, then either the metadata block was tampered with or it was signed by a different identity.

IMO, this improvement is not worth the cost of increasing metadata block size and, thus, egress traffic for Trustless Gateways.

Do you have any other use case for the signature in your mind?

IMO, the clients making retrieval requests don't need this signature for validating the metadata block, as they can rely on guarantees provided by the underlying transport - HTTPS.

> * HTTP Gateways have no concept of "server ED25519" introduced here. 

Good point. We don't require all Gateways to sign the metadata block, SPARK needs the signature only from Storage Providers' servers handling retrieval (`booster-http`).

Let's update the spec to explicitly mention the signature is an optional field.

> How one verifies the signature without knowing the pubkey?
>   * One way to avoid being prescriptive about key type or its location, is to have `sig_key` with CID-encoded public  `libp2p-key` that can be used for signature verification.

As I wrote above, if you don't know the expected server identity, then the signature is not useful for you.

Having said that, I like the idea of adding more details about the identity/public key to the spec. 

The proposed format _CID-encoded public `libp2p-key`_ seems like a good candidate, although AFAICT, that's not the format advertised to IPNI. In IPNI, I see identities in the format that can be used in multiaddr's `/p2p/{id}` part:

```
12D3KooWAWHEbCQy22d45mKbKSewoB1xksDDhR7o5S4mDrSNKXNk
12D3KooWAy5kaLtHf5uS7PZVLjSYd8sGqJ6fn7bxMjqLLZ1uULp9
12D3KooWEiPRcfjXJVehty8okJGJpBZP8zM5UBoCK5yw2MXfx98x
12D3KooWFpv7LP1MUmjfQ8sAUXgJXG5FRMJLnqnJyR32fVboqspB
12D3KooWHKeaNCnYByQUMS2n5PAZ1KZ9xKXqsb4bhpxVJ6bBJg5V
12D3KooWNHwmwNRkMEP6VqDCpjSZkqripoJgN7eWruvXXqC2kG9f
12D3KooWSfsqUahHLCmiENT8oN4FkVtz5pSCxKtNEb7wrR1rrRjk
```

> * If you sign JSON, you want it to be deterministic variant like DAG-JSON, otherwise someone will run into bugs when they use less strict JSON library in different languages.

Makes sense; I'll update the spec to require the metadata to be a DAG-JSON.

    },
    "required": []
  }
}
```

The properties object can include any fields that the server would like to implement. The following JSON schema explicitly mentions certain properties fields in order to reach a convention on their definition as they have existing use cases.

```json
{
  "type": "object",
  "properties": {
    "car_bytes": {
      "description": "The total byte length of the CAR stream (excluding the 0x00 byte and the metadata block)",
      "type": "integer"
    },
    "data_bytes": {
      "description": "Total byte length of the flat file before it was encoded into a CAR file",
      "type": "integer"

>     },

Reviewer: @bajtos what happens when returned CAR is for:
- HAMT-sharded UnixFS directory?
- a single file under some sub-path of HAMT-sharded UnixFS directory?

Is the semantic meaning here to be "raw bytes of all files, ignoring UnixFS directory metadata", or something else?

Author: Great questions! TBH, I don't know the answers. We don't need `data_bytes` for SPARK. I think this field was added based on the discussion in this proposal, but I could not find the specific comment requesting it.

I am proposing to remove `data_field` from the spec. We can introduce it later if there is a clear need. We will better understand the desired semantics at that point.

    "block_count": {
      "description": "Total number of blocks present in the CAR stream (excluding the 0x00 byte and the metadata block, but including duplicates when present)",
      "type": "integer"
    },
    "car_cid": {
      "description": "A hash of the CAR stream giving a CIDv1 with 0x0202 codec",
      "type": "string"
    },
    "b3checksum": {
      "description": "A Blake3 hash (checksum) of the CAR stream (excluding the 0x00 byte and the metadata block). The value should be serialized as a multihash with multibase prefix, preferably using Base58 encoding.",
      "type": "string"

>     },

Reviewer: nit: Hardcoding Blake3 in field name and description makes no sense if you use Multihash.
Rename to `car_checksum` ?

Author: Here is your description of `car_cid`, see https://github.com/ipfs/specs/pull/431#discussion_r1327713175:

> I think it would be also ok to have a documented convention for passing a hash of the CAR stream (aka CAR CID) – maybe name it `car_cid` and use CIDv1 with 0x0202 codec – this convention is already used by .storage folks, no need to invent anything new.

Regarding `b3checksum`:

For SPARK, we specifically need the Blake3 hash of the CAR stream, and we need gateways to always return this hash. In particular, clients cannot ask the server to use Blake3 for the CAR checksum because the server could use this information to detect SPARK clients vs. other clients and provide different quality of service.

I agree it's confusing to have both `car_cid` and `b3checksum`, but I don't see a better solution. Do you?

    "content_path": {
      "description": "The url path in the request as executed by the gateway, e.g. `/ipfs/bafy1234/cat.jpg`. The query string MUST BE stripped from the path.",
      "type": "string"

>     },

Author: Discussion point leading back to https://github.com/ipfs/specs/pull/431#discussion_r1287474957:

How do we represent the information about what content was requested?
- The CID
- An optional path to a file inside UnixFS

the raw data is already present in the parent block that links to the identity
CID.

## CAR `meta` (content type parameter)

The `meta` parameter allows clients to request the server to include additional metadata about the
CAR to be included at the end of the response body.

This parameter can be used with `version=1` only.

When the parameter is not set, the server must not add any extra CAR blocks to the response.

The metadata block is a regular CAR block with the following properties:

- CID specifies multicodec `car-metada` (0x04ff), see
  [multicodec#334](https://github.com/multiformats/multicodec/pull/334).

- The payload contains metadata encoded as DAG-CBOR.

The metadata MUST include the following fields:

- `len` - byte length of the CAR data (excluding the metadata block)
- `b3h` - Blake3 hash (checksum) of the CAR data (excluding the metadata block).
- `b3h_sig` - A signature over `<len><b3h><request>` using server's Ed2559 identity.
  - `len` is encoded as `varint`,
  - `b3h` is encoded as 32 bytes,

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Reviewer 2: @lidel what is the reasoning behind splitting up `dag_params` and `car_params`? Could we instead go for `content_path` and `query_params` to keep it simple and generic (allowing for other query params)?

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Reviewer 2: I agree that the Spark use case belongs in Userland section. However, the individual keys of the metadata section, and what the client must do to implement them, feels like something that should be in the trustless gateway spec. 

Keys like `car_bytes`, `data_bytes`, `block_count` will be used by Spark but also may be used by others, and the definition of a key must be the same for each use case. E.g. if one client sets `data_bytes` to be the total byte length of blocks and another sets it to be the total byte length of the CAR stream then trustless gateway implementers will need to implement different logic for each use case.

What's more, for the Spark use case, we do not want gateway operators to know that they are serving a Spark request and not some other request. Since the Spark ones will be incentivised and other request may not be, servers can simply provide a good retrieval service to Spark clients and a poor service to other clients.

`car_bytes`, `data_bytes`, `block_count` seem generic enough. The troublesome one is then the Blake3Hash and signature.

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Reviewer 2: Perhaps what we need to do is leave this IPIP to concern the metadata block being appended without any constraints on what can be included in it. Then in a separate place, we define a canonical way to include a key value   object in the metadata block and how the server should implement certain useful keys such as `car_bytes` et al

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Reviewer: I think it would be ok to suggest a key name convention for generic things like  `car_bytes` `data_bytes` and `block_count` in the section that described `meta` parameter, as long it is
- scoped to JSON (perhaps list under explicit `meta=eof[+json]`?)
- change requirement from MUST to SHOULD (convention, not a hard requirement)

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Reviewer: I believe that was a conscious design choice, to avoid mixing data selector with details of the transport format (not everything is in URL query params):

- `dag_params` are about what user data was selected, not tied to any specific transport (could be applied to something other than CAR)
   - these are things that land in URL query
- `car_params` are specific to CAR container format, they do not change the user data that was selected, only the way it is represented when sent as CAR  
   -  these are things that land in Accept/Content-Type headers

@patrickwoodhead  that being said, if you want to simplify, this IPIP could go with a single dag-json map named `response_params` (to avoid confusion with URL query params, and account for the fact that server may ignore some of request params when producing a response)

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Author: How about using `retrieval_params` instead of `response_params`? In my mind, response parameters don't describe "what user data was selected".

What is our motivation in SPARK: 
We want the gateway to describe what exactly the client is retrieving (CID, subpath, dag params, car params) and provide a signature over that.

If two SPARK checker clients submit a metadata block with the same retrieval parameters (CID, subpath, dag params, car params) then we want:
- To be able to verify that the clients retrieved the DAG (sub)tree they were expected to retrieve.
- Confidence that both clients made the same request to the gateway (semantically?) and were supposed to receive exactly the same response from the gateway.

> `content_path` - requested content path

When the client requests `GET /ipfs/bafy1234/cat.jpg`, what is `content_path`?
- `/ipfs/bafy1234/cat.jpg`
- `bafy1234/cat.jpg`
- `/cat.jpg`
- something else?

We need the metadata block to describe both the CID requested (`bafy1234`) and the resource subpath if specified (`/cat.jpg`).

>   - The effective query as executed by the gateway. This query is the request url - path and query string arguments.

Author: > I think it would be also ok to have a documented convention for passing a hash of the CAR stream (aka CAR CID) – maybe name it `car_cid` and use CIDv1 with 0x0202 codec – this convention is already used by .storage folks, no need to invent anything new.

+1 to document `car_cid`.

For SPARK, we specifically want a Blake3 hash so that we can use inclusion proofs. That's why we want to use a dedicated field `b3checksum` instead of a more generic `car_cid`.

the raw data is already present in the parent block that links to the identity
CID.

## CAR `meta` (content type parameter)

The `meta=eof` parameter allows clients to request the server to include additional metadata about the

> CAR to be included at the end of the response body.

Reviewer: :thought_balloon: 

If this IPIP ends up doing `<CARv1> <0x00 byte> <dag-cbor-metadata>` would it be less mysterious and more self-explanatory if we did `eof=cbor`?

> CAR to be included at the end of the response body.

Reviewer: :thought_balloon: since the metadata is now outside and after the CAR stream, and @rvagg noted preference for JSON over CBOR, would it be good time to rename this to more self-explanatory `eof=json`? This way we could also support `eof=nil` (removing the need for sending  `{}` when metadata is not needed)

Author: How about `meta=json`?

Author: > `eof=nil` at some point (removing the need for sending  `{}` when metadata is not needed)

FWIW, if we don't need metadata, then we can omit the entire `meta` (or `eof`) parameter.

Reviewer 2: @bajtos what is your reasoning for preferring `meta` over `eof`?

Author: I interpret `meta=json` as "CAR blocks with additional metadata in JSON format". It's open to extension, e.g. allowing metadata blocks in more places than just EOF.

In one of the discussions (sadly not captured in writing), somebody was objecting against using `eof` in `meta=eof` and suggesting to use just `meta` key with no value. I don't remember exact reasons.

On the second thought, considering that we are looking for a short-term fix with least resistance, maybe the key name (`meta` vs `eof`) doesn't matter that much.

Reviewer 2: I like having `json` or `cbor` or other being the value since it makes the metadata options extensible. This means we can start with dag-json as per @rvagg 's suggestion.  For the key, I think that both `eof` and `meta` are not quite right to describe this. Metadata could be placed elsewhere and eof doesnt clear signal that this is metadata. What about `eofmeta=json`?

Author: > What about `eofmeta=json`?

Sounds good to me.

Maybe instead of combining `eof` and `meta`, we can use the name `trailer`? E.g. `trailer=json`

Author: We are going forward with `meta=eof+json`

This parameter SHOULD only be used with CAR `version=1`.
Values other than `eof` SHOULD be ignored.

When the parameter is not set, the server must not add any extra CAR blocks to the response.

> The metadata block is a regular CAR block with the following properties:
> 
> - CID specifies multicodec `car-metadata` (`0x04ff`), see
>   [multicodec#334](https://github.com/multiformats/multicodec/pull/334).

Reviewer: :information_source:  flagging that this custom codec is not a hard requirement for this IPIP.

As noted in "Option 1" from https://github.com/ipfs/specs/pull/431#issuecomment-1670958141  
`meta=eof` could use regular `dag-cbor` in CID it will still be ok, because `eof` behavior  is opt-in via `Accept` request header and there is explicit signaling in `Content-Type` response header returned before CAR stream.

One could argue this is more friendly to existing implementations, as `dag-cbor` code will not create errors in alternative implementations, while a custom code might.

```suggestion
The metadata block is a regular CAR block with the following properties:

- CID specifies multicodec `dag-cbor` (`0x71`)
```
